packages/component/src/Activity/Avatar.js

```diff
@@ -1,24 +1,66 @@
 import classNames from 'classnames';
+import memoize from 'memoize-one';
 import React from 'react';
 
 import connectToWebChat from '../connectToWebChat';
 
-export default connectToWebChat(
-  ({ styleSet }) => ({ styleSet })
-)(
+const connectAvatar = (...selectors) => connectToWebChat(
   ({
-    children,
-    className,
-    fromUser,
-    styleSet
-  }) =>
-    <div
-      className={ classNames(
-        styleSet.avatar + '',
-        { 'from-user': fromUser },
-        (className || '') + ''
-      ) }
-    >
-      { children }
-    </div>
-)
+    styleSet: {
+      options: {
+        botAvatarImage,
+        botAvatarInitials,
+        userAvatarImage,
+        userAvatarInitials
+      }
+    }
+  }, { fromUser }) => ({
+    avatarImage: fromUser ? userAvatarImage : botAvatarImage,
+    avatarInitials: fromUser ? userAvatarInitials : botAvatarInitials
+  }),
+  ...selectors
+);
+
+class Avatar extends React.Component {
+  constructor() {
+    super();
+
+    this.createBackgroundImageStyle = memoize(backgroundImage => ({ backgroundImage: `url(${ encodeURI(backgroundImage) })` }));
```

Comment:
Presumably this is to avoid over-fetching the asset on re-renders, but could it be solved implementing `shouldComponentUpdate` or perhaps extended from `React.PureComponent`?

Each message creates a _new_ avatar, and this memoization would be scoped to that instance of the component, right?

Reply:
- SCU: I prefer to avoid it because it makes the whole app difficult to debug if things goes wrong
- PureComponent: this one works in this scenario. But I am trying to avoid `PureComponent` most of the time. This is because we always expect "children" to be rich content, and `PureComponent` do not have content when to update them, introducing problems similar to SCU

I am leaning towards stateless component approach. [React Hooks](https://reactjs.org/docs/hooks-intro.html) and [`useMemo`](https://reactjs.org/docs/hooks-reference.html#usememo) will fix these issues. But since they are still in preview I am not very keen on moving to it yet, but I will prepare the code so they can be easily translated over.

Reply:
Chiming in here to offer a different approach after discussing with @compulim. I believe that memoization is the correct optimization, however it should be hoisted outside the instance of `Avatar`.

If we define the memoized function outside of the class definition, we will get the benefits across all instances of the avatar component instead of a single one.

Reply:
Yes that what what I was getting at @a-b-r-o-w-n, glad you both worked it out.

Reply:
Your reasons to avoid SCU and PureComponent make great sense. I'm trying to understand the goals of the memoization in this instance a bit better. Beyond avoiding evaluation of `encodeURI` more times than needed, is there another benefit you are trying to attain?

Reply:
If I use `style={{ backgroundImage: ... }}`, props equality check will cause a wasted render (virtual DOM side). I don't worry too much about `encodeURI` perf.

So here, we are trading CPU vs. memory.

@a-b-r-o-w-n `<Avatar>` has 2 upstreamers `<StackedLayout>` and `<CarouselFilmStrip>`, so if we keep the "style" object right there and pass it into `prop`, the code looks weird to me. So I lean to remove memoize for now until we figure out a better memoization story.

Reply:
Moved to our component called `<CroppedImage>`. So we don't need `style` here anymore.